File: app/src/main/java/net/tjado/usbgadget/GadgetFragment.java
package net.tjado.usbgadget;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.ViewFlipper;

import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import java.util.ArrayList;

public class GadgetFragment extends Fragment {

    private GadgetViewModel gadgetViewModel;
    private ViewFlipper rootFlipper;
    private GadgetObject gadget;
    private String[] functionProfileList;
    private View view;

    public GadgetFragment(GadgetObject g) {
        this.gadget = g;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_gadget, container, false);
        gadgetViewModel = new ViewModelProvider(getActivity()).get(GadgetViewModel.class);

        rootFlipper = view.findViewById(R.id.flipper);

        gadgetViewModel.getGadgets().observe(getViewLifecycleOwner(), item -> {
            for (GadgetObject g : item) {
                if (g.getValue("gadget_path").equals(this.gadget.getValue("gadget_path"))) {
                    this.gadget = g;
                    refreshData();
                }
            }
        });

        gadgetViewModel.hasRootPermissions().observe(getViewLifecycleOwner(), item -> {
            rootFlipper.setDisplayedChild(item ? 0 : 1);
        });

        GadgetAssetProfiles gap = new GadgetAssetProfiles(getActivity().getApplication());
        functionProfileList = gap.getAllFunctionProfiles();

        Button addFunction = view.findViewById(R.id.btn_function_add);
        addFunction.setOnClickListener(v -> {
            final AlertDialog.Builder alertBuilder = new AlertDialog.Builder(getContext());
            alertBuilder.setTitle("Add Function to Gadget");
            alertBuilder.setCancelable(true);

            alertBuilder.setItems(functionProfileList, (dialog, which) -> {
                gadgetViewModel.loadFunctionProfile(gadget, functionProfileList[which]);
            });

            alertBuilder.show();
        });

        
        refreshData();

        return view;
    }

    private void refreshData() {
        View v = view;
        TextView manufacturer = v.findViewById(R.id.tv_gadget_manufacturer);
        TextView product = v.findViewById(R.id.tv_gadget_product);
        TextView serialnumber = v.findViewById(R.id.tv_gadget_sn);
        TextView udc = v.findViewById(R.id.tv_gadget_udc);
        TextView path = v.findViewById(R.id.tv_gadget_path);
        Switch status = v.findViewById(R.id.tv_gadget_status);

        manufacturer.setText(gadget.getValue("manufacturer"));
        product.setText(gadget.getValue("product"));
        serialnumber.setText(gadget.getValue("serialnumber"));
        path.setText(gadget.getValue("gadget_path"));
        udc.setText(gadget.getValue("udc"));

        status.setChecked(gadget.isActivated());
        status.setOnClickListener((buttonView) -> {
            if(status.isChecked()) {
                gadget.activate(response -> {
                    gadgetViewModel.updateGadgetData();
                });
            } else {
                gadget.deactivate(response -> {
                    gadgetViewModel.updateGadgetData();
                });
            }
        });

        LinearLayout list = v.findViewById(R.id.list_functons);
        list.removeAllViews();

        ArrayList<String> functions = gadget.getFunctions();
        for(String functionName : functions) {
            View vi = getLayoutInflater().inflate(R.layout.row_function, null);

            TextView tv = vi.findViewById(R.id.name);
            tv.setText(functionName);

            Switch f_status = vi.findViewById(R.id.status);
            f_status.setChecked( gadget.getActiveFunctions().contains(functionName) );
            f_status.setOnClickListener((buttonView) -> {
                if (f_status.isChecked()){
                    gadget.activateFunction(functionName, false);
                } else {
                    gadget.deactivateFunction(functionName, true);
                }
                gadgetViewModel.reloadGadgetData();
            });

            list.addView(vi);
        }
    }
}

File: app/src/main/java/net/tjado/usbgadget/GadgetFragment.kt
package net.tjado.usbgadget


import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle

import android.view.View
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.ViewModelProvider
import androidx.fragment.app.Fragment
import net.tjado.usbgadget.databinding.FragmentGadgetBinding
import net.tjado.usbgadget.databinding.RowFunctionBinding


class GadgetFragment(private var gadget: GadgetObject) : Fragment() {
    private var _binding: FragmentGadgetBinding? = null
    private val binding get() = _binding!!

    private lateinit var gadgetViewModel: GadgetViewModel
    private lateinit var functionProfileList: Array<String>

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentGadgetBinding.inflate(inflater, container, false)
        val view = binding.root

        gadgetViewModel = ViewModelProvider(requireActivity()).get(GadgetViewModel::class.java)
        gadgetViewModel.gadgets.observe(viewLifecycleOwner) { item: List<GadgetObject> ->
            for (g in item) {
                if (g.getValue("gadget_path") == gadget.getValue("gadget_path")) {
                    gadget = g
                    refreshData()
                }
            }
        }

        gadgetViewModel.hasRootPermissions().observe(viewLifecycleOwner) {
                item -> binding.flipper.displayedChild = if (item) 0 else 1
        }

        val assets = GadgetAssetProfiles(requireActivity().application)
        functionProfileList = assets.allFunctionProfiles

        val addFunction = view.findViewById<Button>(R.id.btn_function_add)
        addFunction.setOnClickListener {
            AlertDialog.Builder(requireContext())
                .setTitle("Add Function to Gadget")
                .setCancelable(true)
                .setItems(functionProfileList) { _, which ->
                    gadgetViewModel.loadFunctionProfile(
                        gadget,
                        functionProfileList[which]
                    )
                }.show()
        }

        refreshData()
        return view
    }

    private fun refreshData() {
        binding.tvGadgetManufacturer.text = gadget.getValue("manufacturer")
        binding.tvGadgetProduct.text = gadget.getValue("product")
        binding.tvGadgetSn.text = gadget.getValue("serialnumber")
        binding.tvGadgetPath.text = gadget.getValue("gadget_path")
        binding.tvGadgetUdc.text = gadget.getValue("udc")
        binding.tvGadgetStatus.isChecked = gadget.isActivated
        binding.tvGadgetStatus.setOnClickListener {
            if (binding.tvGadgetStatus.isChecked) {
                gadget.activate { gadgetViewModel.updateGadgetData() }
            } else {
                gadget.deactivate { gadgetViewModel.updateGadgetData() }
            }
        }

        binding.listFunctons.removeAllViews()

        val functions = gadget.functions
        for (functionName in functions) {
            val rowBinding = RowFunctionBinding.inflate(layoutInflater)
            rowBinding.name.text = functionName

            rowBinding.status.isChecked = gadget.activeFunctions.contains(functionName)
            rowBinding.status.setOnClickListener {
                if (rowBinding.status.isChecked) {
                    gadget.activateFunction(functionName, false)
                } else {
                    gadget.deactivateFunction(functionName, true)
                }
                gadgetViewModel.reloadGadgetData()
            }

            binding.listFunctons.addView(rowBinding.root)
        }
    }
}